Update the method of creating an empty file with right size when saving binary files

jiumao2 opened the pull request:
The current method of creating an empty file with `file_size_bytes` is significantly slow, especially for large files. This PR addresses the issue by introducing a more efficient approach.

Current Implementation in `write_binary_recording` (located in `src/spikeinterface/core/recording_tools.py`):
```python
file = open(file_path, "wb+")
file.truncate(file_size_bytes)
file.close()
```  

Proposed Improvement:
```python
file = open(file_path, "wb+")
file.seek(file_size_bytes - 1)
file.write(b'\0')
file.close()
```

Creating a 100 GB empty file on SSD:
```
Time using seek: 0.05744457244873047 seconds
Time using truncate: 88.06437873840332 seconds
```

zm711 commented:
This is cool. Just tagging myself so I can follow this story as Heberto reviews it.

jiumao2 commented:
I’m working with Windows 10 and Python 3.10. Could the issue be related to the operating system?

h-mayorquin commented:
@zm711 could you try the profile that I wrote above?

h-mayorquin commented:
For your convenience:

```python
%%timeit

file_size_bytes = 10 * 1024 * 1024 * 1024
file = open("temporary.bin", "wb+")
file.truncate(file_size_bytes)
file.close()


%%timeit
file_size_bytes = 10 * 1024 * 1024 * 1024
file = open("temporary.bin", "wb+")
file.seek(file_size_bytes - 1)
file.write(b'\0')
file.close()
```

zm711 commented:
Sure. Running to meeting, but right after :)

zm711 commented:
![image](https://github.com/user-attachments/assets/719600c6-8786-4af1-8c1b-599790794db1)
![image](https://github.com/user-attachments/assets/004cba4a-6418-4d18-ab48-4a6045e4b759)

zm711 commented:
This could be a windows thing. So maybe the better strategy is actually per OS. I've been telling my lab mates to deal with it, but it is brutally slow waiting for large binary files to write for KS and MS5.

zm711 commented:
I'll test on my Mac too and report that here as well.

zm711 commented:
<img width="546" alt="Screenshot 2024-09-13 at 1 20 09 PM" src="https://github.com/user-attachments/assets/de050815-d42f-499b-8431-00c0570d357d">

zm711 commented:
Just to be fair I'm also testing python 3.11 on Windows now, but it'll take a little while.

h-mayorquin commented:
@zm711 
If you can test the 100 GiB version that would be great as well.

Yet another place where our linux bias shows up.  I am confused on why seeking would be faster on windows but it should be easy to make this procedure os based as the difference in time for the @jiumao2  is indeed substantial.

zm711 commented:
Sure I'll start with my Mac and do windows later.

h-mayorquin commented:
Whenever you can, man, no rush : ) 
Thanks for helping!

zm711 commented:
Times are exactly the same for 100 on Mac.

zm711 commented:
To be honest running the timeit on Windows is causing my computer to freak out... I don't know what was happening for the truncate. It worked instantly for seek.....

jiumao2 commented:
@zm711 Great speed in testing!

h-mayorquin commented:
Maybe this is related to the zero filling?

https://docs.python.org/3/library/io.html#io.IOBase.truncate

zm711 commented:
The 100 GiB test failed on my computer due to lack of storage. (I've only got a small SSD so this test might be hard for me to do).

zm711 commented on src/spikeinterface/core/recording_tools.py:
```diff
@@ -131,8 +131,10 @@ def write_binary_recording(
+        # create a file with file_size_bytes
         file = open(file_path, "wb+")
-        file.truncate(file_size_bytes)
+        file.seek(file_size_bytes - 1)
```
```suggestion
        if platform.system() == 'Windows':
            file.seek(file_size_bytes - 1)
            file.write(b"\0")
        else:
            file.truncate(file_size_bytes)
```

We could do something like this. Then we would just need to add `import platform` at the top.

h-mayorquin replied:
No, this would work, did you check on mac, it should posix as well but if you have the chance it would cost nothing.

zm711 replied:
Let me label my checks so we can go back easier and see what I did :) 

Mac was faster than linux. Windows was brutally slow.

zm711 replied:
So to summarize

seek-write is 100x faster on Windows
truncate is 2x faster on Mac and Linux.

h-mayorquin replied:
Yeah, I would like to find out what's up but this is a great speed up! I say let's add your suggestion.

zm711 commented on src/spikeinterface/core/recording_tools.py:
```diff
@@ -131,8 +131,10 @@ def write_binary_recording(
         file = open(file_path, "wb+")
-        file.truncate(file_size_bytes)
+        file.seek(file_size_bytes - 1)
+        file.write(b"\0")
```
```suggestion
```

zm711 reviewed (commented):
This would be my suggestion unless you have a better idea Heberto.

jiumao2 commented:
@h-mayorquin I agree that Python on Windows does zero-filling in `truncate`. It will be slower on Windows with larger files and slower SSDs/HHDs. It took me hours to write a 2TB file on my HHD, and it ultimately failed due to memory issues after creating the empty file...

Besides, since `seek` and `write` are both O(1) operations across systems, it might be less meaningful to differentiate the systems for sub-millisecond speed up?

zm711 commented:
> Besides, since seek and write are both O(1) operations across systems, it might be less meaningful to differentiate the systems for sub-millisecond speed up?

I think this is a fair point but at least for me it's less about differentiating for a sub ms speedup and more about keeping the default for posix systems and then optimizing for Windows systems. Until someone feels comfortable knowing exactly why there is this difference (maybe zero filling but the docs aren't clear on this) we could keep our original implementation for posix and the speedup for Windows. Maybe you have some clearer docs on the situation so we can better understand the differences across OSs? Either way we should probably add a comment (either explaining why we added the Windows difference) or a comment why we changed everything to seek-write.

h-mayorquin commented:
> @h-mayorquin I agree that Python on Windows does zero-filling in `truncate`. It will be slower on Windows with larger files and slower SSDs/HHDs. It took me hours to write a 2TB file on my HHD, and it ultimately failed due to memory issues after creating the empty file...
> 
> Besides, since `seek` and `write` are both O(1) operations across systems, it might be less meaningful to differentiate the systems for sub-millisecond speed up?

This is a good point, maybe it doesn't make sense to differentiate.

jiumao2 commented:
> I think this is a fair point but at least for me it's less about differentiating for a sub ms speedup and more about keeping the default for posix systems and then optimizing for Windows systems. Until someone feels comfortable knowing exactly why there is this difference (maybe zero filling but the docs aren't clear on this) we could keep our original implementation for posix and the speedup for Windows. Maybe you have some clearer docs on the situation so we can better understand the differences across OSs? Either way we should probably add a comment (either explaining why we added the Windows difference) or a comment why we changed everything to seek-write.

I'm primarily a Windows user. It is totally up to you how to implement it for posix systems.

zm711 commented:
> I'm primarily a Windows user. It is totally up to you how to implement it for posix systems.

I'm between Windows and Mac so I think this Windows update will be wonderful. I could go either way on the posix implementation vs taking a minuscule slow down to reduce complexity. Could you add a comment just linking back to this PR where we did the profiling to indicate why we needed to switch :) for Windows. If you want to revert back to just one implementation that's fine with me. But I do want a comment so we know why we are doing what we are doing for this really low level operation.

jiumao2 commented:
@zm711 I reverted it back. And I found `with open` fits better here. Is that comment okay?

zm711 reviewed (commented):
This works for me. context manager seems safer too. Thanks @jiumao2.

h-mayorquin reviewed (approved):
Thanks for bringing this to our attention. This is a great contribution that would improve the lives of many window users. 

Thanks @zm711 for reviewing this as well!

zm711 reviewed (approved):
Good by Heberto good by me.

h-mayorquin commented:
My vote is for waiting on this one.
